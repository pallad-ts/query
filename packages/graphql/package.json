{
  "name": "@pallad/query-graphql",
  "version": "2.0.1",
  "description": "Helper for graphql to create @pallad/query related types",
  "main": "compiled/index.js",
  "types": "compiled/index.d.ts",
  "scripts": {
    "test": "pallad-scripts test",
    "lint": "pallad-scripts lint",
    "compile": "pallad-scripts compile",
    "prepublishOnly": "npm run compile"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/pallad-ts/query.git",
    "directory": "packages/main"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "Łukasz Kużyński <lukasz.kuzynski@gmail.com> (http://wookieb.pl/)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pallad-ts/query/issues"
  },
  "homepage": "https://github.com/pallad-ts/query#readme",
  "peerDependencies": {
    "graphql": "^16.3.0",
    "graphql-compose": "^9.0.8"
  },
  "dependencies": {
    "@pallad/builder": "^1.0.1",
    "@pallad/query": "^2.0.0",
    "predicates": "^2.0.3"
  }
}
