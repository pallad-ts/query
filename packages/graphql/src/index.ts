export * from './types';
export * from './createInputSortType';
export * from './createResultListType';
export * from './createSortFieldType';
export * from './createSortTypeBase';
export * from './GraphQLQueryBuilder';
