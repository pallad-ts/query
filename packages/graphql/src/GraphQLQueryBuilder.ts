import {Builder as _Builder} from '@pallad/builder';
import {InputTypeComposer, ObjectTypeComposer, ResolverResolveParams} from "graphql-compose";
import {GraphQLEnumType, GraphQLInputObjectType, GraphQLInt, GraphQLList, GraphQLNonNull, GraphQLObjectType, GraphQLString} from "graphql";
import {Query, Result, SortableFieldDefinition} from "@pallad/query";
import {
	ObjectTypeComposerFieldConfigAsObjectDefinition,
	ObjectTypeComposerFieldConfigMapDefinition
} from "graphql-compose/lib/ObjectTypeComposer";
import {ThunkWithSchemaComposer} from "graphql-compose/lib/utils/definitions";
import {InputTypeComposerFieldConfigMapDefinition} from "graphql-compose/lib/InputTypeComposer";
import {GraphQLSortDirection} from "./types";
import * as is from 'predicates'

const assertEntityTypeIsObjectTypeComposer = is.assert(is.instanceOf(ObjectTypeComposer), 'Entity type must be a type of ObjectTypeComposer');

export class GraphQLQueryBuilder<TEntityType, TQuery extends Query<any>, TContext, TSortableField extends string> extends _Builder {
	private paginationType?: 'byCursor' | 'byOffset';
	private sortableType?: 'multi' | 'single';
	private extraMetaFields?: ObjectTypeComposerFieldConfigMapDefinition<TEntityType, TContext>;
	private sortableFields?: string[];
	private defaultSorting?: SortableFieldDefinition<TSortableField> | Array<SortableFieldDefinition<TSortableField>>;

	constructor(
		private fetcher: (query: TQuery, context: TContext) => Result<TEntityType>,
		private options: GraphQLQueryBuilder.Options<TEntityType, TContext>) {
		super();
	}

	getName() {
		if (this.options.name) {
			return this.options.name;
		}

		return this.options.entityType instanceof GraphQLObjectType ? this.options.entityType.name : this.options.entityType.getTypeName();
	}

	paginateByCursor(): this {
		this.paginationType = 'byCursor'
		return this;
	}

	paginateByOffset(): this {
		this.paginationType = 'byOffset';
		return this;
	}

	multiSortable(sortableFields: TSortableField[]): this {
		this.sortableType = 'multi';
		this.sortableFields = sortableFields;
		return this;
	}

	defaultSortOrder(order: SortableFieldDefinition<TSortableField> | Array<SortableFieldDefinition<TSortableField>>): this {
		this.defaultSorting = order;
		return this;
	}

	singleSortable(sortableFields: TSortableField[]): this {
		this.sortableType = 'single';
		this.sortableFields = sortableFields;
		return this;
	}

	useExtraMetaFields(extraMetaFields: ObjectTypeComposerFieldConfigMapDefinition<TEntityType, TContext>): this {
		this.extraMetaFields = extraMetaFields;
		return this;
	}

	private getResultMetaType() {
		const metaFields: ObjectTypeComposerFieldConfigMapDefinition<any, TContext> = {};
		const metaPaginationFields = this.getMetaPaginationFields();
		if (metaPaginationFields) {
			Object.assign(metaFields, metaPaginationFields);
		}

		if (this.extraMetaFields) {
			Object.assign(metaFields, this.extraMetaFields);
		}

		if (!Object.keys(metaFields).length) {
			return;
		}

		return ObjectTypeComposer.createTemp<any, TContext>({
			name: `${this.getName()}_Query_Result_Meta`,
			fields: metaFields
		});
	}

	private getMetaPaginationFields() {
		if (!this.paginationType) {
			return;
		}

		if (this.paginationType === 'byCursor') {
			return {
				nextPage: {type: GraphQLString},
				previousPage: {type: GraphQLString},
				limit: {type: new GraphQLNonNull(GraphQLInt)}
			};
		}

		if (this.paginationType === 'byOffset') {
			return {
				limit: {type: new GraphQLNonNull(GraphQLInt)},
				offset: {type: GraphQLInt}
			}
		}
	}

	getResultType(): ObjectTypeComposer<Result<TEntityType>, TContext> {
		const metaType = this.getResultMetaType();
		return ObjectTypeComposer.createTemp<Result<TEntityType>, TContext>({
			name: `${this.getName()}_Query_Result`,
			fields: {
				results: {type: this.getEntityTypePlural()},
				...(metaType ? {meta: {type: metaType}} : {})
			}
		});
	}

	private getEntityTypePlural() {
		const entityType = this.options.entityType;
		if (entityType instanceof GraphQLObjectType) {
			return new GraphQLList(new GraphQLNonNull(entityType));
		}
		return entityType.getTypePlural();
	}

	getQueryArgType() {
		const fields: ThunkWithSchemaComposer<InputTypeComposerFieldConfigMapDefinition, any> = {
			filters: {type: this.options.filtersType}
		};

		const sortByType = this.getSortByInputType();
		if (sortByType) {
			fields.sortBy = {
				type: sortByType
			};
		}

		if (this.paginationType) {
			fields.limit = {type: GraphQLInt};
			if (this.paginationType === 'byOffset') {
				fields.offset = {type: GraphQLInt};
			} else if (this.paginationType === 'byCursor') {
				fields.after = {type: GraphQLString};
				fields.before = {type: GraphQLString}
			}
		}

		return InputTypeComposer.createTemp<TContext>({
			name: this.getName() + '_Query',
			fields
		});
	}

	private getSortByInputType() {
		if (!this.sortableFields) {
			return;
		}
		const baseType = InputTypeComposer.createTemp({
			name: this.getName() + '_Query_Sort',
			fields: {
				direction: {type: GraphQLSortDirection},
				field: {
					type: new GraphQLEnumType({
						name: this.getName() + '_Query_Sort_Field',
						values: this.sortableFields!.reduce((result, field) => {
							result[field] = {value: field};
							return result;
						}, {} as Record<string, { value: string }>)
					})
				}
			}
		});

		if (this.sortableType === 'single') {
			return baseType;
		}
		return baseType.getTypePlural();
	}

	private getFiltersType() {
		const filtersType = this.options.filtersType;
		if (filtersType instanceof GraphQLInputObjectType) {
			return new GraphQLNonNull(filtersType);
		}

		return filtersType.getTypeNonNull();
	}

	getQueryField(): ObjectTypeComposerFieldConfigAsObjectDefinition<unknown, TContext, { query: TQuery }> {
		this.assertConfigurationSanity();
		return {
			type: this.getResultType(),
			args: {
				query: {
					type: this.getQueryArgType()
				}
			},
			resolve: (source, args, context) => {
				let query = args.query;
				if (!query) {
					query = {filters: {}} as TQuery;
				}
				if (!query.filters) {
					query.filters = {};
				}
				return this.fetcher(query, context);
			}
		}
	}

	private assertConfigurationSanity() {
		this.assertDefaultSortingType();
	}

	private assertDefaultSortingType() {
		if (Array.isArray(this.defaultSorting) && this.sortableType === 'single') {
			throw new Error('Cannot apply default sort order on multiple fields while query was configured for single sort order only');
		}
	}

	asResolver(name: string = 'query') {
		this.assertResolverConfig();

		const entityType = this.options.entityType as ObjectTypeComposer;
		const queryField = this.getQueryField();

		entityType.addResolver({
			name,
			args: queryField.args,
			type: queryField.type,

			resolve: (rp: ResolverResolveParams<any, any>) => {
				return queryField.resolve!(rp.source, rp.args, rp.context, rp.info);
			}
		});

		return entityType.getResolver(name);
	}

	private assertResolverConfig() {
		assertEntityTypeIsObjectTypeComposer(this.options.entityType);
	}
}

export namespace GraphQLQueryBuilder {
	export interface Options<TEntityType, TContext> {
		name?: string;
		entityType: ObjectTypeComposer<TEntityType, TContext> | GraphQLObjectType<TEntityType, TContext>
		filtersType: GraphQLInputObjectType | InputTypeComposer<TContext>;
	}
}
