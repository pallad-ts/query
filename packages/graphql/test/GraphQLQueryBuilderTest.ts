import { ObjectTypeComposer, printSchemaComposer, SchemaComposer } from "graphql-compose";
import { GraphQLID, GraphQLInputObjectType, GraphQLSchema, GraphQLString } from "graphql";
import { GraphQLQueryBuilder } from "@src/GraphQLQueryBuilder";
import { QueryDescriptor } from "@pallad/query-descriptor";
import { DEFAULT_SORTING, SORTABLE_FIELDS } from "./fixtures";
import * as sinon from "sinon";
import { parse, execute } from "graphql";
import { fromZodValidation } from "@pallad/violations-zod";
import { z } from "zod";

function printSchemaForBuilder(queryBuilder: GraphQLQueryBuilder<any, any, any>) {
	const composer = new SchemaComposer();
	composer.Query.addFields({
		query: queryBuilder.getQueryField(),
	});

	return printSchemaComposer(buildSchemaForBuilder(queryBuilder), {
		sortTypes: true,
	});
}

function buildSchemaForBuilder(queryBuilder: GraphQLQueryBuilder<any, any, any>) {
	const composer = new SchemaComposer();
	composer.Query.addFields({
		query: queryBuilder.getQueryField(),
	});
	return composer;
}

function runQueryForBuilder(
	queryBuilder: GraphQLQueryBuilder<any, any, any>,
	query: string,
	variables: Record<string, any>,
	context: any
) {
	const composer = buildSchemaForBuilder(queryBuilder);
	return runQuery(composer.buildSchema(), query, variables, context);
}

function runQuery(
	schema: GraphQLSchema,
	query: string,
	variables: Record<string, any>,
	context: any
) {
	return execute({
		contextValue: context,
		schema,
		document: parse(query),
		variableValues: variables,
	});
}

describe("GraphQLQueryBuilder", () => {
	const entityType = ObjectTypeComposer.createTemp({
		name: "Entity",
		fields: {
			id: { type: GraphQLID },
			name: { type: GraphQLString },
		},
	});

	const filtersType = new GraphQLInputObjectType({
		name: "Filters",
		fields: {
			test: { type: GraphQLString },
		},
	});

	const CONTEXT = { some: "context" };

	describe("simple", () => {
		const builder = new GraphQLQueryBuilder(new QueryDescriptor(), {
			entityType,
			fetcher() {
				return {} as any;
			},
			filtersType,
		});

		it("schema", () => {
			expect(printSchemaForBuilder(builder)).toMatchSnapshot();
		});
	});

	describe("cursor pagination", () => {
		const builder = new GraphQLQueryBuilder(
			new QueryDescriptor()
				.cursorPagination({ defaultLimit: 100, maxLimit: 1000 })
				.singleSorting(SORTABLE_FIELDS.slice(), DEFAULT_SORTING),
			{
				entityType,
				fetcher() {
					return {} as any;
				},
				filtersType,
			}
		);

		it("schema", () => {
			expect(printSchemaForBuilder(builder)).toMatchSnapshot();
		});

		describe("resolving query", () => {
			const QUERY = `
				query ($query: Entity_Query!) {
					query (query: $query) {
						results {
							id
							name
						}
					}
				}
			`;
			it("success for simple", async () => {
				const stub = sinon.stub().resolves({
					results: [],
				});

				const builder = new GraphQLQueryBuilder(new QueryDescriptor(), {
					entityType,
					fetcher: stub,
					filtersType,
				});

				const result = await runQueryForBuilder(
					builder,
					QUERY,
					{ query: { filters: {} } },
					CONTEXT
				);

				expect(result.errors).toBeUndefined();

				sinon.assert.calledWithExactly(
					stub,
					{
						filters: {},
					},
					CONTEXT
				);
			});

			it("fails if violates filters schema", async () => {
				const stub = sinon.stub().resolves({
					results: [],
				});
				const builder = new GraphQLQueryBuilder(
					new QueryDescriptor().filtersValidator(input => {
						return fromZodValidation(
							z.object({
								test: z.enum(["foo", "bar"]),
							}),
							input
						);
					}),
					{
						entityType,
						fetcher: stub,
						filtersType,
					}
				);

				const result = await runQueryForBuilder(
					builder,
					QUERY,
					{
						query: {
							filters: {
								test: "zee",
							},
						},
					},
					CONTEXT
				);

				expect(result.errors).not.toBeUndefined();
				expect(result.errors).toMatchSnapshot();
			});
		});
	});

	describe("offset pagination", () => {
		const builder = new GraphQLQueryBuilder(
			new QueryDescriptor().offsetPagination({ defaultLimit: 100, maxLimit: 1000 }),
			{
				entityType,
				fetcher() {
					return {} as any;
				},
				filtersType,
			}
		);

		it("schema", () => {
			expect(printSchemaForBuilder(builder)).toMatchSnapshot();
		});
	});

	describe("multi sorting", () => {
		const builder = new GraphQLQueryBuilder(
			new QueryDescriptor().multiSorting(SORTABLE_FIELDS.slice(), [DEFAULT_SORTING]),
			{
				entityType,
				fetcher() {
					return {} as any;
				},
				filtersType,
			}
		);

		it("schema", () => {
			expect(printSchemaForBuilder(builder)).toMatchSnapshot();
		});
	});

	describe("single sorting", () => {
		const builder = new GraphQLQueryBuilder(
			new QueryDescriptor().singleSorting(SORTABLE_FIELDS.slice(), DEFAULT_SORTING),
			{
				entityType,
				fetcher() {
					return {} as any;
				},
				filtersType,
			}
		);

		it("schema", () => {
			expect(printSchemaForBuilder(builder)).toMatchSnapshot();
		});
	});

	it("as resolver", async () => {
		const entityType = ObjectTypeComposer.createTemp({
			name: "Entity",
			fields: {
				id: { type: GraphQLID },
				name: { type: GraphQLString },
			},
		});

		const builder = new GraphQLQueryBuilder(
			new QueryDescriptor()
				.cursorPagination({ defaultLimit: 100, maxLimit: 1000 })
				.singleSorting(SORTABLE_FIELDS.slice(), DEFAULT_SORTING),
			{
				entityType,
				fetcher() {
					return {
						results: [{ id: "1", name: "foo" }],
					};
				},
				filtersType,
			}
		);

		builder.attachQueryResolverToEntity();

		const composer = new SchemaComposer();
		composer.Query.addFields({
			query: entityType.getResolver("query"),
		});

		const result = await runQuery(
			composer.buildSchema(),
			`
				query ($query: Entity_Query!) {
					query (query: $query) {
						results {
							id
							name
						}
					}
				}
			`,
			{ query: { filters: {} } },
			CONTEXT
		);

		expect(result.errors).toBeUndefined();
		expect(result.data).toEqual({
			query: {
				results: [{ id: "1", name: "foo" }],
			},
		});
	});
});
