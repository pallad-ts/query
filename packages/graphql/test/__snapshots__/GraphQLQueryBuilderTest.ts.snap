// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLQueryBuilder cursor pagination resolving query fails if violates filters schema 1`] = `
[
  [GraphQLError: Invalid query],
]
`;

exports[`GraphQLQueryBuilder cursor pagination schema 1`] = `
"type Entity {
  id: ID
  name: String
}

input Entity_Query {
  filters: Filters
  sortBy: Entity_Query_Sort_Input
  after: String
  before: String
  limit: PositiveInt
}

input Entity_Query_Sort_Input {
  direction: SortDirection!
  field: Entity_Sort_Field!
}

type Entity_Result {
  results: [Entity!]!
  meta: Entity_Result_Meta!
}

type Entity_Result_Meta {
  nextPage: String
  previousPage: String
  limit: PositiveInt!
  sortBy: Entity_Result_Sort!
}

type Entity_Result_Sort {
  direction: SortDirection!
  field: Entity_Sort_Field!
}

enum Entity_Sort_Field {
  foo
  bar
}

enum Entity_Sort_Field {
  foo
  bar
}

input Filters {
  test: String
}

"""
The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`"4"\`) or integer (such as \`4\`) input value will be accepted as an ID.
"""
scalar ID

"""Integers that will have a value greater than 0."""
scalar PositiveInt

"""Integers that will have a value greater than 0."""
scalar PositiveInt

type Query {
  query(query: Entity_Query): Entity_Result!
}

enum SortDirection {
  ASC
  DESC
}

enum SortDirection {
  ASC
  DESC
}

"""
The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.
"""
scalar String

"""
The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.
"""
scalar String

"""
The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.
"""
scalar String"
`;

exports[`GraphQLQueryBuilder multi sorting schema 1`] = `
"type Entity {
  id: ID
  name: String
}

input Entity_Query {
  filters: Filters
  sortBy: [Entity_Query_Sort_Input!]
}

input Entity_Query_Sort_Input {
  direction: SortDirection!
  field: Entity_Sort_Field!
}

type Entity_Result {
  results: [Entity!]!
  meta: Entity_Result_Meta!
}

type Entity_Result_Meta {
  sortBy: [Entity_Result_Sort!]!
}

type Entity_Result_Sort {
  direction: SortDirection!
  field: Entity_Sort_Field!
}

enum Entity_Sort_Field {
  foo
  bar
}

enum Entity_Sort_Field {
  foo
  bar
}

input Filters {
  test: String
}

"""
The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`"4"\`) or integer (such as \`4\`) input value will be accepted as an ID.
"""
scalar ID

type Query {
  query(query: Entity_Query): Entity_Result!
}

enum SortDirection {
  ASC
  DESC
}

enum SortDirection {
  ASC
  DESC
}

"""
The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.
"""
scalar String

"""
The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.
"""
scalar String"
`;

exports[`GraphQLQueryBuilder offset pagination schema 1`] = `
"type Entity {
  id: ID
  name: String
}

input Entity_Query {
  filters: Filters
  limit: PositiveInt
  offset: PositiveInt
}

type Entity_Result {
  results: [Entity!]!
  meta: Entity_Result_Meta!
}

type Entity_Result_Meta {
  limit: PositiveInt!
  offset: PositiveInt!
}

input Filters {
  test: String
}

"""
The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`"4"\`) or integer (such as \`4\`) input value will be accepted as an ID.
"""
scalar ID

"""Integers that will have a value greater than 0."""
scalar PositiveInt

"""Integers that will have a value greater than 0."""
scalar PositiveInt

type Query {
  query(query: Entity_Query): Entity_Result!
}

"""
The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.
"""
scalar String

"""
The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.
"""
scalar String"
`;

exports[`GraphQLQueryBuilder simple schema 1`] = `
"type Entity {
  id: ID
  name: String
}

input Entity_Query {
  filters: Filters
}

type Entity_Result {
  results: [Entity!]!
}

input Filters {
  test: String
}

"""
The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`"4"\`) or integer (such as \`4\`) input value will be accepted as an ID.
"""
scalar ID

type Query {
  query(query: Entity_Query): Entity_Result!
}

"""
The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.
"""
scalar String

"""
The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.
"""
scalar String"
`;

exports[`GraphQLQueryBuilder single sorting schema 1`] = `
"type Entity {
  id: ID
  name: String
}

input Entity_Query {
  filters: Filters
  sortBy: Entity_Query_Sort_Input
}

input Entity_Query_Sort_Input {
  direction: SortDirection!
  field: Entity_Sort_Field!
}

type Entity_Result {
  results: [Entity!]!
  meta: Entity_Result_Meta!
}

type Entity_Result_Meta {
  sortBy: Entity_Result_Sort!
}

type Entity_Result_Sort {
  direction: SortDirection!
  field: Entity_Sort_Field!
}

enum Entity_Sort_Field {
  foo
  bar
}

enum Entity_Sort_Field {
  foo
  bar
}

input Filters {
  test: String
}

"""
The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`"4"\`) or integer (such as \`4\`) input value will be accepted as an ID.
"""
scalar ID

type Query {
  query(query: Entity_Query): Entity_Result!
}

enum SortDirection {
  ASC
  DESC
}

enum SortDirection {
  ASC
  DESC
}

"""
The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.
"""
scalar String

"""
The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.
"""
scalar String"
`;
